### ISSUE TITLE
False positive undefined-variable when variable name in decoration matches function argument

### ISSUE THREAD

**Issue by mthuurne** [PyCQA/pylint#3791]
### Steps to reproduce

Run pylint on the following code:

```py
def decorate(seq):
    return lambda f: f

@decorate(x*x for x in range(3))
def f(x):
    pass

@decorate(x*x for x in range(3))
def g(y):
    pass
```

### Current behavior

pylint reports:
```
4:12: E0602: Undefined variable 'x' (undefined-variable)
```
Line 4 being the `@decorate` preceding `f()`.

Note that no error is reported for `g()`: the problem only occurs if the variable name inside the comprehension matches one of the decorated function's arguments. Which is strange, since they should be in separate scopes.

### Expected behavior

No error is reported.

### pylint --version output

```
pylint 2.6.0
astroid 2.4.2
Python 3.8.4 (default, Jul 14 2020, 09:08:56) [GCC]
```

This problem did not occur until I upgraded from pylint 2.5.3 to 2.6.0.

**Comment by stpierre** [comment]
We're hitting this as well. The other requirement of this is that the variable needs to be used *twice* in a list comprehension in the decorator. (This is really remarkably specific!) So for instance, these all give the false-positive `undefined-variable`:

```
@decorate(i * i for i in [])
def thing1(i):
    pass

@decorate(dict(i=i, j=i) for i in [])
def thing2(i):
    pass

@decorate((i, i) for i in [])
def thing3(i):
    pass
```

But this does not:

```
@decorate(i for i in [])
def thing4(i):
    pass
```

Neither do other ways of declaring and introducing new variables, for instance lambdas. These are both fine:

```
@decorate(lambda i: i * i)
def thing5(i):
    pass

@decorate(lambda i: [i * i for i in []])
def thing6(i):
    pass
```

This is also broken in pylint 2.7.

**Comment by rr-** [comment]
I believe this is relevant place to patch
`pylint/checkers/variables.py:992`

```python3
            # if the name node is used as a function default argument's value or as
            # a decorator, then start from the parent frame of the function instead
            # of the function frame - and thus open an inner class scope
            if (
                current_consumer.scope_type == "function"
                and self._defined_in_function_definition(node, current_consumer.node)
            ):
                # ignore function scope if is an annotation/default/decorator, as not in the body
                continue
```
